Improve data structure, validation and performance

ManfredKarrer opened the pull request:
Fixes https://github.com/bisq-network/bisq/issues/2554
Fixes https://github.com/bisq-network/bisq/issues/2555
Fixes https://github.com/bisq-network/bisq/issues/2545
Fixes https://github.com/bisq-network/bisq/issues/2539
Fixes https://github.com/bisq-network/bisq/issues/2415

Adding new params and bonded roles is possible after adding the UNDEFINED enums and avoid throwing exceptions if enum is not found.

ManfredKarrer committed:
Only log if hashchain has been updated

ManfredKarrer committed:
Fix incorrect maybeUpdateHashChain call

ManfredKarrer committed:
Add dao-testnet explorer, update testnet url

ManfredKarrer committed:
Add UNDEFINED enum entry to all dao enums to allow updates

For backward compatibility we resolve an unknown new enum entry with
UNDEFINED if UNDEFINED is available in the enum.
Adding a bonded role and adding a param is tested with that and works.
Other changes need to be tested before implemented as they might have
more consequences for consensus.

ManfredKarrer committed:
Log time needed for rpc getBlock call

ManfredKarrer committed:
Merge branch 'master' into misc-dao-improvements

ManfredKarrer committed:
Add comment

ManfredKarrer committed:
Adjust THRESHOLD_CHANGE_PARAM and THRESHOLD_CONFISCATION

ManfredKarrer committed:
Rename PRICE_NODE_OPERATOR to DATA_RELAY_NODE_OPERATOR

The price node will likely be used for further data relay tasks and the
delivery of the fee is already not well convered by the price term...

ManfredKarrer committed:
Add BTC_DONATION_ADDRESS_OWNER

ManfredKarrer committed:
Add BTC_DONATION_ADDRESS_OWNER translation string

ManfredKarrer committed:
Fix incorrect maxIncrease value

ManfredKarrer committed:
Rename getMiningFeeAndTxSize to getBlindVoteMiningFeeAndTxSize

ManfredKarrer committed:
Make ARBITRATOR_FEE % based, Add more validations

ManfredKarrer committed:
Refactor ChangeParamValidator

ManfredKarrer committed:
Refactor: Rename ValidationException to ProposalValidationException

ManfredKarrer committed:
Use ProposalValidatorProvider to operate on concrete ProposalValidator

We used ProposalValidator for most validation processes but that missed
the custom validation in the sub classed for each proposal type.
ProposalValidator is now abstract and ProposalValidatorProvider returns
instance matching to proposal type.

ManfredKarrer committed:
Use block height of proposal tx if available

To be able to apply the validation also on past cycles we need to use
the block height if the proposal tx and not the current one. Just in
case the tx is not confirmed (when temp proposal gets published) we use
the current height, but that would anyway match the cycle.
In a past cycle params could have been different and validation need to
use the correct param value from that cycle.

ManfredKarrer committed:
Merge branch 'master' into misc-dao-improvements

ManfredKarrer committed:
Change blind vote phase and vote reveal durations

Blind vote phase changed from 4 days to 3 days and vote reveal from
2 days to 3 days.

ManfredKarrer committed:
Change min/max factor for phase durations from 3 to 2

ManfredKarrer committed:
Increase bond for youtube operator

ManfredKarrer committed:
Add 3 new roles, adjust bond amount

ManfredKarrer committed:
Fix comments

ManfredKarrer committed:
Change bond lock time from 60 days to 75 days

ManfredKarrer committed:
Remove requiredQuorum and requiredThreshold from EvaluatedProposal

We would break the hash chain if we would change the quorum or
threshold values in the param enum.

ManfredKarrer committed:
Store requiredBond and unlockTime in RoleProposal

ManfredKarrer committed:
Add extraDataMap to BlindVote and Proposal

In case we need to add additional data in future we are can use that map
without breaking consensus or breaking the hash chain.

ManfredKarrer commented:
Sorry this PR becomes much bigger as expected. ;-( But as there are some breaking changes, I still prefer to have it in one and with one test session.

ManfredKarrer committed:
Add ConsensusCritical to proposal validators

Changing validation rules can potentially break consensus (e.g. a past
proposal has been accepted according to old rules, and might become
invalid by new rules. The result of the proposal would become
invalidated  then.

devinbileck commented on core/src/main/java/bisq/core/dao/monitoring/ProposalStateMonitoringService.java:
```diff
@@ -134,15 +134,18 @@ public void onDaoStateChanged(Block block) {
             long ts = System.currentTimeMillis();
             for (int i = genesisBlockHeight; i < blockHeight; i++) {
-                maybeUpdateHashChain(i);
+                hashChainUpdated = hashChainUpdated || maybeUpdateHashChain(i);
```
Does this not have the unintended consequence of only updating the hashchain for the first cycle? Since once maybeUpdateHashChain returns true (first block of blind vote phase), this statement will short-circuit and not call maybeUpdateHashChain for any other blocks?

devinbileck replied:
Never mind. Just saw your subsequent commit :)

devinbileck commented on core/src/main/java/bisq/core/dao/monitoring/ProposalStateMonitoringService.java:
```diff
@@ -134,15 +134,18 @@ public void onDaoStateChanged(Block block) {
+                hashChainUpdated = hashChainUpdated || maybeUpdateHashChain(i);
+            }
+            if (hashChainUpdated) {
+                log.info("updateHashChain for {} items took {} ms",
```
Perhaps could be clarified as blocks rather than items?

ManfredKarrer replied:
Fixed in new commit.

ManfredKarrer committed:
Dont' call updateBsqBalance while batch processing

We got a concurrent modification error when BTC blockchain sync was
running while DAO state was updated.

ManfredKarrer committed:
Change log levels, improve logging

ManfredKarrer committed:
Rename db file to enforce usage of new file with new data structure

devinbileck commented on core/src/main/java/bisq/core/dao/governance/proposal/param/ChangeParamValidator.java:
```diff
@@ -214,24 +222,64 @@ private void validateParamValue(Param param, String inputValue, int chainHeight)
+        }
+    }
+
+    private void validateAddressValue(String currentParamValue, String inputValue) {
```
Should add `throws AddressFormatException` to this method signature

ManfredKarrer replied:
AddressFormatException is a RuntimeException. But agree whould be more clean to add it. Will consider to wrap it into ValidationException.

ManfredKarrer replied:
Used ParamValidationException in last commit.

devinbileck commented on core/src/main/java/bisq/core/dao/governance/proposal/param/ChangeParamValidator.java:
```diff
@@ -214,24 +222,64 @@ private void validateParamValue(Param param, String inputValue, int chainHeight)
+            return Result.SAME;
+
+        if (max == 0)
+            return Result.OK;
```
Are you assuming that if max is zero, then so is min? If max is zero but min is non-zero (not sure if there is a use case for that) and the change is <1, then it wont validate min.

ManfredKarrer replied:
Yes that is not correct. Was a bit too lazy... Basically there is no use case for setting one to 0 and the other not. But should handle it correclty... The whole validation code could need more love...

ManfredKarrer committed:
Add SHA3-256 hash function

SHA3-256 is about 30% faster than SHA256

ManfredKarrer committed:
Stream block batch processing on render frames

To avoid that the UI gets frozen at batch processing of blocks we
delay each parsing to the next render frame. The total parsing time is
just about 5% slower that way but the UI can render updates.
We also changed the hash for the daoState as the hashing of the full
state becomes quite heavy. The size of the blocks is about 1,4 MB for
7000 blocks (dao testnet). As on a new block only the last block in the
chain got added and as we use the previous hash in the hash chain we
do not need to hash the full blocks list but only the last block.
By that we decrease batch processing time from 30 sec to 7 sec. and data
size of the daoState from 1,4 MB to 200 kb.
Also added progress display of missing blocks in the Tx UI.

ManfredKarrer committed:
Fix logging

ManfredKarrer committed:
Use ParamValidationException instead of AddressFormatException

devinbileck commented on core/src/main/resources/i18n/displayStrings.properties:
```diff
@@ -1540,7 +1544,9 @@ dao.bond.bondedRoleType.BTC_NODE_OPERATOR=BTC node operator
 dao.bond.bondedRoleType.MARKETS_OPERATOR=Markets operator
 # suppress inspection "UnusedProperty"
-dao.bond.bondedRoleType.BSQ_EXPLORER_OPERATOR=BSQ explorer operator
+dao.bond.bondedRoleType.BSQ_EXPLORER_OPERATOR=Markets operator
```
I assume this string should not have been changed?

ManfredKarrer replied:
Ups, thanks.

ManfredKarrer committed:
Add UNDEFINED entries

ManfredKarrer committed:
Support exportToJson also for lite node

ManfredKarrer committed:
Remove SHA3-256 hash as it turned out it is slower as SHA256

In some tests it seemed that SHA3-256 is 30% fast than SHA256 but later
with the data we used in dao testnet it was actually slower.
To avoid confusion which hash function to use and to avoid mixing them
without strong reason I prefer to remove it again.

ManfredKarrer committed:
Misc cleanups and refactorings

ManfredKarrer committed:
Update json export for vote results

Try to reflect domain structure and field names.

ManfredKarrer committed:
Add BONDED_ROLE_FACTOR to param to react on BSQ price changes

To avoid the need to change the required bond in the BondedRoleType
if the BSQ price changes we use the BONDED_ROLE_FACTOR param where the
factor can be changed. In the BondedRoleType we use the requiredBondUnit
which will be multiplied with the BONDED_ROLE_FACTOR value to get the
required bond amount.

ManfredKarrer committed:
Cleanup TODOs

ManfredKarrer committed:
Don't support spending of pending BTC utxos in the BSQ wallet.

ManfredKarrer committed:
Remove nonBsqTxOutputMap from daoState

We can evaluate the nonBsqTxOutputs without storing it.
Storing them in the map would have required to remove them as well
once withdrawn from the wallet.

ManfredKarrer committed:
Apply code inspection, cleanup

ManfredKarrer committed:
Fix string

ManfredKarrer commented:
@sqrrm @devinbileck @ripcurlx
I know that is a monster PR and hard to review.
Still would love to get it soon merged so we have some time for dev testing before next release.
I will continue testing the next hours, but tested already quite a lot. If nobody can review in the next hour I will consider to merge despite an ACK. If any NACK arrives I will not merge of course....

ManfredKarrer committed:
Merge branch 'master' into misc-dao-improvements

devinbileck reviewed (approved):
utACK
Overall looks fine to me. However, I am not well experienced in this area of the code and can just make general observations.
Only concern I had was regarding translations for the validation messages. I assume that is still todo and will be completed separately?

ManfredKarrer commented:
Yes that is on my todo list. I try to focus on the critical issue first and keep that for later.

ManfredKarrer committed:
Update db files

ManfredKarrer committed:
Add phaseChangeListener only after batch processing

ManfredKarrer committed:
Update db files

ManfredKarrer committed:
Use translation strings and show min/max values at change param validation

ManfredKarrer committed:
Fix tests

ManfredKarrer committed:
Fix % display

ManfredKarrer committed:
Add missing @EqualsAndHashCode(exclude = {"date"})

Was unintendely removed in latest refactorings

sqrrm commented on core/src/main/java/bisq/core/dao/DaoFacade.java:
```diff
@@ -680,4 +685,27 @@ public double getRequiredThreshold(Proposal proposal) {
         return proposalService.getRequiredQuorum(proposal);
     }
+
+    public long getRequiredBond(Optional<RoleProposal> roleProposal) {
```
I feel these functions should reside in DaoStateService for consistency. This works, most other actual work is done where it's more relevant rather than in DaoFacade itself.

ManfredKarrer replied:
Hm, not sure. Proposal is another domain as DaoState so I prefer to keep it contained in ProposalService. Sure there is much overlap but DaoStateService is already huge so in doubt I prefer to keep it in ProposalService. With DaoFacade I am not sure if we should keep it mid term. Its quite boilerplate.

sqrrm replied:
Just leave it for now. But would be nice to find a better place for it eventually.

sqrrm reviewed (commented):
utACK
